### PR TITLE
[ZEPPELIN-3644] SPARQL-Interpreter support via Apache Jena ARQ

### PR DESCRIPTION
**Pull request by hsteinmueller**
### What is this PR for?
Interpreter for SPARQL-queries on any SPARQL-endpoint via [Apache Jena ARQ](https://jena.apache.org/documentation/query/).

[SPARQL](https://www.w3.org/TR/sparql11-query/) is an RDF query language able to retrieve and manipulate data stored in Resource Description Framework (RDF) format.

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-3644

### How should this be tested?
* Unit tests are provided
* https://travis-ci.org/hsteinmueller/zeppelin

### Screenshots (if appropriate)
sparql-example

### Questions:
* Does the licenses files need update? Yes, added in PR
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes, added in PR

### CODE REVIEW

File: sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+  @Override
+  public void close() {
+  }
+
+  @Override
+  public InterpreterResult interpret(String queryString, InterpreterContext context) {
+    final String serviceEndpoint = getProperty(SPARQL_SERVICE_ENDPOINT);
+    final String replaceURIs = getProperty(SPARQL_REPLACE_URIS);
+    final String removeDatatypes = getProperty(SPARQL_REMOVE_DATATYPES);
```
Comment (zjffdu) [inline sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java]: These variables should be initialized in open method

File: sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        return new InterpreterResult(
+          InterpreterResult.Code.ERROR,
+            "Error: " + e.getMessage());
+      }
+    }
+  }
+
+  @Override
+  public void cancel(InterpreterContext context) {
+  }
```
Comment (zjffdu) [inline sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java]: `cancel` is unable to be supported ? Or do you have plan to support it ?

File: sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java (hunk 1 above)
Comment (hsteinmueller) [reply sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java]: fixed

File: sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java (hunk 2 above)
Comment (hsteinmueller) [reply sparql/src/main/java/org/apache/zeppelin/fuseki/SparqlInterpreter.java]: done: added support for cancel

**Comment by zjffdu** [comment]
Thanks for the contribution @hsteinmueller , this PR LGTM, could you paste the travis build link ?

**Comment by alexott** [comment]
Just a general question - maybe instead of `sparql` name for interpreter, use something like, `sparql.jena`, so it's possible in the future to use different implementations? Or extend implementation to have different backends selected via configuration?

P.S. I'm really to keen to try this implementation...

**Comment by zjffdu** [comment]
IIUC, `jena` is one implementation of `sparql`. Since interpreter text `%sparql` is user facing, I would prefer to configure the implementation of `sparql` in interpreter setting instead of via interpreter text.

**Comment by hsteinmueller** [comment]
> Thanks for the contribution @hsteinmueller , this PR LGTM, could you paste the travis build link ?

https://travis-ci.org/hsteinmueller/zeppelin

> IIUC, `jena` is one implementation of `sparql`. Since interpreter text `%sparql` is user facing, I would prefer to configure the implementation of `sparql` in interpreter setting instead of via interpreter text.

Do you want me to update the interpreter settings with this functionality or will this be changed once another implementation gets added?

**Comment by zjffdu** [comment]
@hsteinmueller Could you update the pom.xml to resolve the conflict ?

**Comment by zjffdu** [comment]
> > Thanks for the contribution @hsteinmueller , this PR LGTM, could you paste the travis build link ?
> 
> https://travis-ci.org/hsteinmueller/zeppelin
> 
> > IIUC, `jena` is one implementation of `sparql`. Since interpreter text `%sparql` is user facing, I would prefer to configure the implementation of `sparql` in interpreter setting instead of via interpreter text.
> 
> Do you want me to update the interpreter settings with this functionality or will this be changed once another implementation gets added?

I mean to use interpreter setting to choose which implementation.  e.g. Add interpreter setting `sparql.type`.  IIUC, user don't care about the implementation of sparql ( I may be wrong ), only administrator care about that, so we should put ti in interpreter setting which is often maintained by administrator.

**Comment by hsteinmueller** [comment]
> @hsteinmueller Could you update the pom.xml to resolve the conflict ?

done


> > > Thanks for the contribution @hsteinmueller , this PR LGTM, could you paste the travis build link ?
> > 
> > 
> > https://travis-ci.org/hsteinmueller/zeppelin
> > > IIUC, `jena` is one implementation of `sparql`. Since interpreter text `%sparql` is user facing, I would prefer to configure the implementation of `sparql` in interpreter setting instead of via interpreter text.
> > 
> > 
> > Do you want me to update the interpreter settings with this functionality or will this be changed once another implementation gets added?
> 
> I mean to use interpreter setting to choose which implementation. e.g. Add interpreter setting `sparql.type`. IIUC, user don't care about the implementation of sparql ( I may be wrong ), only administrator care about that, so we should put ti in interpreter setting which is often maintained by administrator.

I understood what you meant, I just wanted to know if this is needed right now or if we add this setting in the future when another implementation is needed.
Nonetheless I tried adding support for different implementations. I hope this works for you ☺️

**Comment by zjffdu** [comment]
I think it is fine to add this setting until there's another implementation.

**Comment by zjffdu** [comment]
LGTM, will merge if no more comment

**Pull request closed without merging**